docs/proposals/memory-management.md

```diff
@@ -0,0 +1,135 @@
+
+2. Burstable - Memory limits guaranteed up to user request
+
+3. Best Effort - No guaranty - memory availability can change over time.
+
+The scheduler does not take into account the total usage on the node. 
+Swapping is turned off by default. Kubelet does not track or manage swap files or partitions.
+Kubelet (via cadvisor) can track the memory usage of system daemons and pods. It does not take any action based on usage. 
+
+The kernel OOM killer is expected to be enabled. oom_score_adj is used to protect guaranteed jobs and system daemons. Burstable jobs are given a oom_score_adj that will result in them having a oom_score of `1000` as soon as they hit their original request, making them a likely candidate for being killed by the oom killer. Best effort tasks always have a `oom_score_adj` of `1000` which makes them the most likely candidates to be killed. The kernel usually chooses the highest memory consuming process ([along with a few other policies](https://lwn.net/Articles/317814/))
+
+####Issues with Current model
+The life of a node is blissful up until it has memory pressure, at which points the kernel resorts to killing processes.
+
+* Memory is inherently overcommitted.
+
+ This combined with lack of swap space, increases the probability of system OOMs. This slows down the responsiveness of the system and could have secondary effects like high tail latencies for system daemons.
+
+* A container in a pod with Restart Always policy will be restarted continuously even if it OOMs. 
+
+* Other than the Guaranteed class, jobs in other classes don't have a memory limit.
+
+* Any process in a container can be killed due to either system memory pressure or exceeding container limits.
+
+ There is no container cleanup policy in the event of an OOM kill. What should happen to a pod if a process in one of its container's was OOM killed? 
+
+* Kubelet cannot associate a system memory pressure OOM kill to a specific container or Pod. 
+
+* Memory usage of system daemons is unrestricted. This results in dynamic memory availability on the node. 
+
+* All burstable jobs that exceed their request is equally likely to be killed. Ideally, the job that exceeds the most should be killed. 
+
+* There are no limits placed on kernel resources consumed by jobs (proc, fd, sockets, network buffers, etc)
+
+#####Ideas for improving memory management
+
+1. Define OOM handling policy for containers and Pods:
+
+    *Suggestion*: Whenever an OOM kill occurs in a container, mark that entire container as OOM killed and terminate it. We can provide a knob to alter this behavior if required
+    If a Pod has an app container in "guaranteed" class and a "side-car" container in "Burstable" class, what happens to the pod when the "side-car" is OOM killed? Will the Pod continue to run without the "side-car"?
+    Requiring all containers in a Pod to be of the same QoS class will be simpler.
+
+2. Impose limits on "best effort" and "Burstable" jobs
```

Comment:
Is this related to the idea of having a user-space control loop that keeps the cgroup limit for best effort and burstable slightly above usage, so that it can be killed when it is close to causing system OOM, instead of causing system OOM?